File: internal/ui/theme.go
package ui

import (
	"image/color"
	"strconv"
	"strings"

	"ctail/internal/config"
)

// Colors holds the resolved NRGBA colors for the current theme.
type Colors struct {
	BgPrimary     color.NRGBA
	BgSecondary   color.NRGBA
	BgSurface     color.NRGBA
	BgHover       color.NRGBA
	TextPrimary   color.NRGBA
	TextSecondary color.NRGBA
	TextMuted     color.NRGBA
	Accent        color.NRGBA
	AccentHover   color.NRGBA
	Border        color.NRGBA
	Danger        color.NRGBA
	Success       color.NRGBA
	Warning       color.NRGBA
	TabActive     color.NRGBA
	TabInactive   color.NRGBA
	BadgeColor    color.NRGBA
	ScrollTrack   color.NRGBA
	ScrollThumb   color.NRGBA
}

// DefaultColors returns Catppuccin Mocha dark colors.
func DefaultColors() Colors {
	return Colors{
		BgPrimary:     hexColor("#1e1e2e"),
		BgSecondary:   hexColor("#181825"),
		BgSurface:     hexColor("#313244"),
		BgHover:       hexColor("#45475a"),
		TextPrimary:   hexColor("#cdd6f4"),
		TextSecondary: hexColor("#a6adc8"),
		TextMuted:     hexColor("#6c7086"),
		Accent:        hexColor("#89b4fa"),
		AccentHover:   hexColor("#74c7ec"),
		Border:        hexColor("#45475a"),
		Danger:        hexColor("#f38ba8"),
		Success:       hexColor("#a6e3a1"),
		Warning:       hexColor("#f9e2af"),
		TabActive:     hexColor("#1e1e2e"),
		TabInactive:   hexColor("#181825"),
		BadgeColor:    hexColor("#f9e2af"),
		ScrollTrack:   hexColor("#181825"),
		ScrollThumb:   hexColor("#45475a"),
	}
}

// ColorsFromTheme converts a config.ThemeColors to Colors.
func ColorsFromTheme(tc config.ThemeColors) Colors {
	return Colors{
		BgPrimary:     hexColor(tc.BgPrimary),
		BgSecondary:   hexColor(tc.BgSecondary),
		BgSurface:     hexColor(tc.BgSurface),
		BgHover:       hexColor(tc.BgHover),
		TextPrimary:   hexColor(tc.TextPrimary),
		TextSecondary: hexColor(tc.TextSecondary),
		TextMuted:     hexColor(tc.TextMuted),
		Accent:        hexColor(tc.Accent),
		AccentHover:   hexColor(tc.AccentHover),
		Border:        hexColor(tc.Border),
		Danger:        hexColor(tc.Danger),
		Success:       hexColor(tc.Success),
		Warning:       hexColor(tc.Warning),
		TabActive:     hexColor(tc.TabActive),
		TabInactive:   hexColor(tc.TabInactive),
		BadgeColor:    hexColor(tc.BadgeColor),
		ScrollTrack:   hexColor(tc.ScrollTrack),
		ScrollThumb:   hexColor(tc.ScrollThumb),
	}
}

// hexColor parses a hex color string like "#1e1e2e" to NRGBA.
func hexColor(hex string) color.NRGBA {
	hex = strings.TrimPrefix(hex, "#")
	if len(hex) == 3 {
		hex = string([]byte{hex[0], hex[0], hex[1], hex[1], hex[2], hex[2]})
	}
	if len(hex) != 6 {
		return color.NRGBA{A: 255}
	}
	r, _ := strconv.ParseUint(hex[0:2], 16, 8)
	g, _ := strconv.ParseUint(hex[2:4], 16, 8)
	b, _ := strconv.ParseUint(hex[4:6], 16, 8)
	return color.NRGBA{R: uint8(r), G: uint8(g), B: uint8(b), A: 255}
}

// HexToNRGBA is exported for use in highlight rendering.
func HexToNRGBA(hex string) color.NRGBA {
	return hexColor(hex)
}

File: macos/Sources/ctailmac/Theme.swift
import AppKit
import CtailCore

// `Theme` and `ThemeColors` are engine records (hex strings, Go-compatible JSON
// keys); the catalogue and custom-theme loading live in core/src/themes.rs.
// This file layers the AppKit conveniences on top.

extension ThemeColors {
    /// Neutral grey palette used only as a pre-launch placeholder.
    static let placeholder = ThemeColors(
        bgPrimary: "#1e1e1e", bgSecondary: "#181818", bgSurface: "#2a2a2a", bgHover: "#3a3a3a",
        textPrimary: "#e0e0e0", textSecondary: "#b0b0b0", textMuted: "#808080", accent: "#6bcbff",
        accentHover: "#5bb0e0", border: "#3a3a3a", danger: "#ff6b6b", success: "#a6e3a1",
        warning: "#f9e2af", tabActive: "#1e1e1e", tabInactive: "#181818", badgeColor: "#f9e2af",
        scrollbarTrack: "#181818", scrollbarThumb: "#3a3a3a")

    // Convenience NSColors used by the log surface and chrome.
    var background: NSColor { Theme.hex(bgPrimary) }
    var backgroundAlt: NSColor { Theme.hex(bgSecondary) }
    var surface: NSColor { Theme.hex(bgSurface) }
    var hover: NSColor { Theme.hex(bgHover) }
    var foreground: NSColor { Theme.hex(textPrimary) }
    var muted: NSColor { Theme.hex(textMuted) }
    var gutter: NSColor { Theme.hex(textMuted) }
    var selection: NSColor { Theme.hex(bgHover) }
    var accentColor: NSColor { Theme.hex(accent) }
    var borderColor: NSColor { Theme.hex(border) }
    var dangerColor: NSColor { Theme.hex(danger) }
    var successColor: NSColor { Theme.hex(success) }
    var warningColor: NSColor { Theme.hex(warning) }
    var badge: NSColor { Theme.hex(badgeColor) }
}

extension Theme {
    func palette(mode: String) -> ThemeColors { mode == "light" ? light : dark }

    /// NSColor -> "#rrggbb".
    static func hexString(_ color: NSColor) -> String {
        let c = color.usingColorSpace(.sRGB) ?? color
        return String(format: "#%02x%02x%02x",
                      Int((c.redComponent * 255).rounded()),
                      Int((c.greenComponent * 255).rounded()),
                      Int((c.blueComponent * 255).rounded()))
    }

    static func hex(_ s: String) -> NSColor {
        var h = s.trimmingCharacters(in: .whitespaces)
        if h.hasPrefix("#") { h.removeFirst() }
        if h.count == 3 { h = h.map { "\($0)\($0)" }.joined() }   // #abc -> #aabbcc
        var v: UInt64 = 0
        Scanner(string: h).scanHexInt64(&v)
        return NSColor(srgbRed: CGFloat((v >> 16) & 0xff) / 255,
                       green:   CGFloat((v >> 8) & 0xff) / 255,
                       blue:    CGFloat(v & 0xff) / 255, alpha: 1)
    }
}

/// All themes (21 built-ins + user-supplied custom themes from the config dir).
enum ThemeCatalog {
    static var builtIns: [Theme] { builtInThemes() }

    static func all(custom dir: URL? = nil) -> [Theme] { allThemes(customDir: dir?.path) }

    /// Resolves a theme name + mode to a concrete palette, falling back to the
    /// default (Catppuccin dark) if the name is unknown.
    static func palette(name: String, mode: String, custom dir: URL? = nil) -> ThemeColors {
        resolvePalette(name: name, mode: mode, customDir: dir?.path)
    }
}
